File: binary_search/74.search-a-2-d-matrix.py
#
# @lc app=leetcode id=74 lang=python3
#
# [74] Search a 2D Matrix
#

# @lc code=start
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return False
        R, C = len(matrix), len(matrix[0])
        l, r = 0, R*C-1

        while l <= r:
            mid = (l+r)//2
            # TODO: //C and %C here, not R
            num = matrix[mid//C][mid % C]

            if num == target:
                return True
            elif num < target:
                l = mid+1
            else:
                r = mid - 1

        return False


# @lc code=end

s = Solution()
print(s.searchMatrix([[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], 3))

File: binary_search/74.search-a-2-d-matrix.cpp
/*
 * @lc app=leetcode id=74 lang=cpp
 *
 * [74] Search a 2D Matrix
 */

#include "bits/stdc++.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  bool searchMatrix(vector<vector<int>>& matrix, int target) {
    int kR = matrix.size();
    int kC = matrix[0].size();
    int l = 0, r = kR * kC - 1;
    while (l <= r) {
        int m = l + (r - l) / 2;
        // TODO: //C and %C here, not R
        int num = matrix[m / kC][m % kC];

        if (num == target)
            return true;
        else if (num < target)
            l = m + 1;
        else
            r = m - 1;
    }

    return false;
  }
};
// @lc code=end
